### PR TITLE
Don't URL-decode path segments

### PR DESCRIPTION
**Pull request by philandstuff**
Currently, the standard requires implementations to URL-decode unquoted
path segments.  I think this is a mistake.

In particular, the URLs http://example.com/foo:bar and
http://example.com/foo%3Abar are potentially different URLs, even though
`:` is URL-encoded as `%3A`.  They are different because `:` is a
reserved character, which means that servers may interpret the `:` as a
special delimiter rather than the literal colon character.

An example of this kind of processing is the "matrix parameters" in
http://example.com/foo;bar=baz/quux -- matrix parameters are paremeters
on a path component, similar to how a query string contains parameters
on the whole URL.  Matrix parameters use the reserved characters `;` and
`=` as delimiter syntax.  If you want to use these characters within a
matrix parameter name or value, you need to percent-encode them.  This
demonstrates why `;` and `%3B` may not be equivalent in path segments.

By requiring implementations to URL-decode path segments, we
unintentionally fold distinct URLs into the same representation.

This commit fixes that by changing the internal representation to be the
original (encoded) text, and by requiring *quoted* path segments to be
URL-*encoded* to match.  This changes the binary representation for
URIs, but the binary representation for URIs is rarely used because it
cannot appear in a resolved and normalized expression.

Reserved characters are defined here: https://tools.ietf.org/html/rfc3986#section-2.2

Matrix URIs: https://www.w3.org/DesignIssues/MatrixURIs.html

Related to #581 and dhall-lang/dhall-haskell#1109.

### CODE REVIEW

**Comment by philandstuff** [comment]
Corresponding dhall-haskell change: https://github.com/dhall-lang/dhall-haskell/pull/1235

**Comment by singpolyma** [comment]
So, the purity and using-URI-libraries-should-just-work side of me cringes at this, but I do recall for example that del.icio.us URLs back in the day used to treat `/tag/a+b` and `/tag/a%2Bb` differently, even if reason indicates they should be the same.  So this is not just a theoretical concern.  Storing the parts of a URI encoded feels very odd (as you would never store other syntaxes in a partially-parsed state) but I don't have another suggestion for this corner.

File: tests/binary-decode/success/unit/imports/HttpsImportWithTrickyPathCharsA.diag [hunk 1]
```diff
@@ -1 +1 @@
-[24, null, 0, 1, null, "authority", "path/with?funny-characters", null]
+[24, null, 0, 1, null, "authority", "path:with!funny-characters", null]
```
Comment (sjakobi) [inline tests/binary-decode/success/unit/imports/HttpsImportWithTrickyPathCharsA.diag]: What's the reason for this change?

Comment (philandstuff) [reply tests/binary-decode/success/unit/imports/HttpsImportWithTrickyPathCharsA.diag]: Without URL-decoding, theres no longer a way to construct an import such that its binary representation has a path component containing a `/` or `?` character. So this test is testing an impossible scenario. I changed it to a more representative scenario.

**Comment by philandstuff** [comment]
> Storing the parts of a URI encoded feels very odd (as you would never store other syntaxes in a partially-parsed state) but I don't have another suggestion for this corner.

I think the power of a URL is that it's a hierarchical syntax, where different parts of the syntax are for different consumers.  For example, in `http://user@example.com:8080/foo/bar;x=1/baz?y=2#thing`:

 - `http`, `user`, `example.com`, `8080` and `thing` are for the user-agent. The user-agent needs to parse them out and know how to deal with them directly.
 - The user agent sends the rest, unmodified and unparsed, to the `example.com` as a GET request: `GET /foo/bar;x=1/baz?y=2`
 - `example.com` happens to be a cache.  it can use `/foo/bar;x=1/baz?y=2` as a cache key, provided it doesn't assume two different paths are in fact the same. (sometimes you get problems if your caching layer thinks the query string isn't significant when it is)
 - we have a cache miss, so it gets forwarded to `appserver.example.com`.  this is an application server hosting many apps. It knows that the first path component, `foo`, should be handled by FooServlet.  It passes the rest, `/bar;x=1/baz?y=2`, to FooServlet unmodified
 - FooServlet parses the rest.  It knows that we're at path `bar`, with matrix parameter `x` set to `1`, subpath `baz`, with query parameter `y` set to `2`.  It resolves and returns a response.
 - finally, the user-agent finds the identifier `thing` in the returned document.

Note that the user-agent doesn't (and shouldn't) know or care about matrix parameters or query parameters.  This isn't the whole story - the user-agent needs to be able to parse enough of a URL path and query string in order to resolve relative references, so that if it sees a link to `../jim`, it can do the right thing.  The amount of parsing you need to do to resolve relative references is strictly less than the amount of parsing you need to identify query parameters and matrix parameters.  RFC 3986 describes the relative reference resolution algorithm.  RFC 3986 expressly does not define some of the syntax mentioned here, such as query parameters or matrix parameters. Servers are given a lot of leeway to mint URLs using whatever syntax they like (provided they only use reserved characters for creating novel syntax).

In short: I think the whole magic of URLs is that they are hierarchical identifiers, with different actors having responsibility for parsing different parts of them.  We should only partially parse them because that's what we're supposed to do.
